File: apps/desktop/src/platform/desktop-host/common/desktop-handshake.ts
import { z } from 'zod'

export const DESKTOP_HANDSHAKE_PREFIX = 'LOOPAL_DESKTOP ' as const
export const DESKTOP_EVENT_PREFIX = 'LOOPAL_DESKTOP_EVENT ' as const
export const DESKTOP_PROTOCOL_VERSION = 1 as const
export const DESKTOP_TRANSPORT = 'tcp_jsonrpc_ndjson' as const
export const DESKTOP_CAPABILITY_HUB_UI = 'hub_ui_v1' as const
export const DESKTOP_CAPABILITY_WORKSPACE = 'workspace_v1' as const
export const DESKTOP_REQUIRED_CAPABILITIES = [
  DESKTOP_CAPABILITY_HUB_UI,
  DESKTOP_CAPABILITY_WORKSPACE,
] as const

const capabilities = z.array(z.string().min(1)).refine(
  (items) => DESKTOP_REQUIRED_CAPABILITIES.every((item) => items.includes(item)),
  'required desktop capabilities are missing',
)

const common = z.object({
  protocol_version: z.literal(DESKTOP_PROTOCOL_VERSION),
  server_version: z.string().min(1),
  pid: z.number().int().positive(),
  parent_pid: z.number().int().positive().optional(),
})

export const DesktopHandshakeSchema = z.discriminatedUnion('phase', [
  common.extend({
    phase: z.literal('alive'),
    addr: z.string().regex(/^127\.0\.0\.1:\d+$/),
    token: z.string().min(1),
    transport: z.literal(DESKTOP_TRANSPORT),
    capabilities,
  }).strict(),
  common.extend({
    phase: z.literal('ready'),
    session_id: z.string().min(1),
  }).strict(),
  common.extend({
    phase: z.literal('error'),
    code: z.string().min(1),
    message: z.string().min(1),
  }).strict(),
])

export const DesktopSessionCreatedHandshakeSchema = common.extend({
  phase: z.literal('session_created'),
  session_id: z.string().min(1),
}).strict()

export type DesktopCoreHandshake = z.infer<typeof DesktopHandshakeSchema>
export type DesktopSessionCreatedHandshake = z.infer<
  typeof DesktopSessionCreatedHandshakeSchema
>
export type DesktopHandshake = DesktopCoreHandshake | DesktopSessionCreatedHandshake
export type DesktopAliveHandshake = Extract<DesktopCoreHandshake, { phase: 'alive' }>
export type DesktopReadyHandshake = Extract<DesktopCoreHandshake, { phase: 'ready' }>
export type DesktopActivationHandshake = DesktopSessionCreatedHandshake | DesktopReadyHandshake

/** Parse one physical stdout line. Non-protocol logging is deliberately ignored. */
export function parseDesktopHandshakeLine(line: string): DesktopHandshake | undefined {
  const normalized = line.replace(/[\r\n]+$/, '')
  if (normalized.startsWith(DESKTOP_EVENT_PREFIX)) {
    return DesktopSessionCreatedHandshakeSchema.parse(
      JSON.parse(normalized.slice(DESKTOP_EVENT_PREFIX.length)),
    )
  }
  if (!normalized.startsWith(DESKTOP_HANDSHAKE_PREFIX)) return undefined
  return DesktopHandshakeSchema.parse(
    JSON.parse(normalized.slice(DESKTOP_HANDSHAKE_PREFIX.length)),
  )
}

File: crates/loopal-ipc/src/desktop_handshake.rs
//! Versioned startup handshake for `loopal desktop serve`.
//!
//! Core handshake records use `LOOPAL_DESKTOP `. Optional lifecycle events
//! use a separate prefix so older strict Desktop parsers can ignore them.

use serde::{Deserialize, Serialize};

pub const DESKTOP_HANDSHAKE_PREFIX: &str = "LOOPAL_DESKTOP ";
pub const DESKTOP_EVENT_PREFIX: &str = "LOOPAL_DESKTOP_EVENT ";
pub const DESKTOP_PROTOCOL_VERSION: u32 = 1;
pub const DESKTOP_TRANSPORT: &str = "tcp_jsonrpc_ndjson";
pub const DESKTOP_CAPABILITY_HUB_UI: &str = "hub_ui_v1";
pub const DESKTOP_CAPABILITY_WORKSPACE: &str = "workspace_v1";

/// A single machine-readable line emitted while the Desktop Host starts.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DesktopHandshake {
    pub protocol_version: u32,
    pub server_version: String,
    pub pid: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_pid: Option<u32>,
    #[serde(flatten)]
    pub event: DesktopHandshakeEvent,
}

/// Alive lets Desktop attach before the root Agent can request approval;
/// SessionCreated exposes the durable identity before full readiness.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "phase", rename_all = "snake_case")]
pub enum DesktopHandshakeEvent {
    Alive {
        addr: String,
        token: String,
        transport: String,
        capabilities: Vec<String>,
    },
    Ready {
        session_id: String,
    },
    SessionCreated {
        session_id: String,
    },
    Error {
        code: String,
        message: String,
    },
}

impl DesktopHandshake {
    pub fn alive(
        server_version: impl Into<String>,
        pid: u32,
        parent_pid: Option<u32>,
        addr: impl Into<String>,
        token: impl Into<String>,
    ) -> Self {
        Self {
            protocol_version: DESKTOP_PROTOCOL_VERSION,
            server_version: server_version.into(),
            pid,
            parent_pid,
            event: DesktopHandshakeEvent::Alive {
                addr: addr.into(),
                token: token.into(),
                transport: DESKTOP_TRANSPORT.to_string(),
                capabilities: vec![
                    DESKTOP_CAPABILITY_HUB_UI.to_string(),
                    DESKTOP_CAPABILITY_WORKSPACE.to_string(),
                ],
            },
        }
    }

    pub fn ready(
        server_version: impl Into<String>,
        pid: u32,
        parent_pid: Option<u32>,
        session_id: impl Into<String>,
    ) -> Self {
        Self {
            protocol_version: DESKTOP_PROTOCOL_VERSION,
            server_version: server_version.into(),
            pid,
            parent_pid,
            event: DesktopHandshakeEvent::Ready {
                session_id: session_id.into(),
            },
        }
    }

    pub fn session_created(
        server_version: impl Into<String>,
        pid: u32,
        parent_pid: Option<u32>,
        session_id: impl Into<String>,
    ) -> Self {
        Self {
            protocol_version: DESKTOP_PROTOCOL_VERSION,
            server_version: server_version.into(),
            pid,
            parent_pid,
            event: DesktopHandshakeEvent::SessionCreated {
                session_id: session_id.into(),
            },
        }
    }

    pub fn error(
        server_version: impl Into<String>,
        pid: u32,
        parent_pid: Option<u32>,
        code: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self {
            protocol_version: DESKTOP_PROTOCOL_VERSION,
            server_version: server_version.into(),
            pid,
            parent_pid,
            event: DesktopHandshakeEvent::Error {
                code: code.into(),
                message: message.into(),
            },
        }
    }

    /// Encode as exactly one newline-terminated protocol record.
    pub fn encode(&self) -> String {
        let json = serde_json::to_string(self).expect("DesktopHandshake must serialize");
        let prefix = match self.event {
            DesktopHandshakeEvent::SessionCreated { .. } => DESKTOP_EVENT_PREFIX,
            _ => DESKTOP_HANDSHAKE_PREFIX,
        };
        format!("{prefix}{json}\n")
    }

    /// Parse a protocol line. Non-protocol stdout returns `Ok(None)`; a line
    /// with our prefix but malformed JSON returns an error so callers can
    /// distinguish pollution from a broken protocol implementation.
    pub fn parse(line: &str) -> Result<Option<Self>, serde_json::Error> {
        let trimmed = line.trim_end_matches(['\r', '\n']);
        let (json, event_prefix) =
            if let Some(json) = trimmed.strip_prefix(DESKTOP_HANDSHAKE_PREFIX) {
                (json, false)
            } else if let Some(json) = trimmed.strip_prefix(DESKTOP_EVENT_PREFIX) {
                (json, true)
            } else {
                return Ok(None);
            };
        let handshake: Self = serde_json::from_str(json)?;
        let session_created = matches!(
            handshake.event,
            DesktopHandshakeEvent::SessionCreated { .. }
        );
        if event_prefix != session_created {
            return Err(serde_json::Error::io(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                "Desktop handshake phase used the wrong protocol prefix",
            )));
        }
        Ok(Some(handshake))
    }
}

#[cfg(test)]
mod tests;
